Sources/swift-format/Frontend/Frontend.swift

```diff
@@ -0,0 +1,202 @@
+      let sourceData = fileHandle.readDataToEndOfFile()
+      defer { fileHandle.closeFile() }
+      return String(data: sourceData, encoding: .utf8)
+    }()
+
+    init(fileHandle: FileHandle, path: String, configuration: Configuration) {
+      self.fileHandle = fileHandle
+      self.path = path
+      self.configuration = configuration
+    }
+  }
+
+  /// The diagnostic engine to which warnings and errors will be emitted.
+  final let diagnosticEngine: DiagnosticEngine = {
+    let engine = DiagnosticEngine()
+    let consumer = PrintingDiagnosticConsumer()
+    engine.addConsumer(consumer)
+    return engine
+  }()
+
+  /// Options that apply during formatting or linting.
+  final let lintFormatOptions: LintFormatOptions
+
+  /// Loads formatter configuration files.
+  final var configurationLoader = ConfigurationLoader()
+
+  /// Advanced options that are useful for developing/debugging but otherwise not meant for general
+  /// use.
+  final var debugOptions: DebugOptions {
+    [
+      lintFormatOptions.debugDisablePrettyPrint ? .disablePrettyPrint : [],
+      lintFormatOptions.debugDumpTokenStream ? .dumpTokenStream : [],
+    ]
+  }
+
+  /// Indicates whether any errors were emitted during execution.
+  final var errorsWereEmitted: Bool { diagnosticEngine.hasErrors }
+
+  /// Creates a new frontend with the given options.
+  ///
+  /// - Parameter lintFormatOptions: Options that apply during formatting or linting.
+  init(lintFormatOptions: LintFormatOptions) {
+    self.lintFormatOptions = lintFormatOptions
+  }
+
+  /// Runs the linter or formatter over the inputs.
+  final func run() {
+    let paths = lintFormatOptions.paths
+
+    if paths.isEmpty {
+      processStandardInput()
+    } else {
+      processPaths(paths)
+    }
+  }
+
+  /// Called by the frontend to process a single file.
+  ///
+  /// Subclasses must override this method to provide the actual linting or formatting logic.
+  ///
+  /// - Parameter fileToProcess: A `FileToProcess` that contains information about the file to be
+  ///   processed.
+  func processFile(_ fileToProcess: FileToProcess) {
+    fatalError("Must be overridden by subclasses.")
+  }
+
+  /// Processes source content from standard input.
+  private func processStandardInput() {
+    guard let configuration = configuration(
+      atPath: lintFormatOptions.configurationPath,
+      orInferredFromSwiftFileAtPath: nil)
+    else {
+      // Already diagnosed in the called method.
+      return
+    }
+
+    let fileToProcess = FileToProcess(
+      fileHandle: FileHandle.standardInput,
+      path: lintFormatOptions.assumeFilename ?? "<stdin>",
+      configuration: configuration)
+    processFile(fileToProcess)
+  }
+
+  /// Processes source content from a list of files and/or directories provided as paths.
+  private func processPaths(_ paths: [String]) {
+    precondition(
+      !paths.isEmpty,
+      "processFiles() should only be called when paths is non-empty.")
```

Comment:
Should this be `processPaths`?

Reply:
It should be `processPaths(_:)` 😛  That's what I get for trying to have a helpful sanity check.